### PR TITLE
Support setting image data from DirectByteBuffer

### PR DESCRIPTION
**Pull request by westinrm**
Allowing creation of KtxTexture from a DirectByteBuffer lets you interop with images written by other native libraries (for example, GDAL) without having to copy everything to and from the heap.

### CODE REVIEW

**Comment by MarkCallow** [comment]
I am on the verge of releasing 4.1.0, I will wait until that is done before reviewing and merging this.

Thank you for the contribution.

File: interface/java_binding/src/test/java/org/khronos/ktx/test/KtxTexture1Test.java [hunk 1]
```diff
@@ -116,4 +117,21 @@ public void testCreate() {
...
+
+        info.setGlInternalformat(KtxInternalformat.GL_COMPRESSED_RGBA_ASTC_4x4_KHR);
+        info.setBaseWidth(10);
+        info.setBaseHeight(10);
+
+        KtxTexture1 texture = KtxTexture1.create(info, KtxCreateStorage.ALLOC);
+        assertNotNull(texture);
+
+        ByteBuffer imageData = ByteBuffer.allocateDirect(10 * 10);
+        texture.setImageFromBuffer(0, 0, 0, imageData);
```
Comment (MarkCallow) [inline interface/java_binding/src/test/java/org/khronos/ktx/test/KtxTexture1Test.java:133]: You're not testing the result for KTX_SUCCESS.

File: interface/java_binding/src/test/java/org/khronos/ktx/test/KtxTexture1Test.java [hunk 2]
```diff
@@ -116,4 +117,21 @@ public void testCreate() {
...
+        KtxTextureCreateInfo info = new KtxTextureCreateInfo();
+
+        info.setGlInternalformat(KtxInternalformat.GL_COMPRESSED_RGBA_ASTC_4x4_KHR);
+        info.setBaseWidth(10);
+        info.setBaseHeight(10);
+
+        KtxTexture1 texture = KtxTexture1.create(info, KtxCreateStorage.ALLOC);
+        assertNotNull(texture);
+
+        ByteBuffer imageData = ByteBuffer.allocateDirect(10 * 10);
```
Comment (MarkCallow) [inline interface/java_binding/src/test/java/org/khronos/ktx/test/KtxTexture1Test.java:132]: I think you should put some known data in the ByteBuffer and after `setImageFromBuffer`, check that the texture has the expected data.

File: interface/java_binding/src/test/java/org/khronos/ktx/test/KtxTexture2Test.java [hunk 3]
```diff
@@ -227,4 +235,28 @@ public void testCreate() {
...
+        KtxTextureCreateInfo info = new KtxTextureCreateInfo();
+
+        info.setBaseWidth(10);
+        info.setBaseHeight(10);
+        info.setVkFormat(VkFormat.VK_FORMAT_ASTC_4x4_SRGB_BLOCK);
+
+        KtxTexture2 texture = KtxTexture2.create(info, KtxCreateStorage.ALLOC);
+        assertNotNull(texture);
+
+        ByteBuffer imageData = ByteBuffer.allocateDirect(10 * 10);
```
Comment (MarkCallow) [inline interface/java_binding/src/test/java/org/khronos/ktx/test/KtxTexture2Test.java:250]: Same comment as for `texture1Test`.

File: interface/java_binding/src/test/java/org/khronos/ktx/test/KtxTexture2Test.java [hunk 4]
```diff
@@ -227,4 +235,28 @@ public void testCreate() {
...
+        texture.setImageFromBuffer(0, 0, 0, imageData);
```
Comment (MarkCallow) [inline interface/java_binding/src/test/java/org/khronos/ktx/test/KtxTexture2Test.java:251]: Same comment as for `texture1Test`.

**Review by MarkCallow — CHANGES_REQUESTED** [review]
Looks good to me but I'm no Java expert anymore. It's a long time since I was making 3D graphics engines for J2ME.

Please strengthen the tests as indicated.

4.1.0 is still not released. I'm waiting for a new code signing certificate.

**Comment by westinrm** [comment]
I'm going to close this.  It's no longer useful to me and I won't have time to look at it for at least a month anyways.  I'll re-open if it becomes relevant again.

**Pull request closed without merging**